### DISCUSSION TITLE
Linux: Running the app as a webapp (e.g. from a Raspberry Pi)

### DISCUSSION THREAD

**Discussion by jha749** [DeFiCh/app#736] [Ideas]
**Is your feature request related to a problem? Please describe.**
I'm running a node on my Raspberry Pi, but in some cases, would also like a nice GUI to check my wallet. The Wallet app at the moment does not support pointing it to a node on another machine. Running it as a web app on the Pi could be a solution.

**Describe the solution you'd like**
Running the Wallet app as a webapp, served from my Raspberry Pi. I managed to get a proof of concept running, by hardcoding the rpc settings in the code. It would be nice to have a better solution to override the rpc settings. Theoretically, with a good set up, this could also allow pointing the Electron version of the app to a node running on another machine.

The main point here is to discuss a good way to override the rpc settings.

**Describe alternatives you've considered**
n/a

**Additional context**
Originating reddit post: https://www.reddit.com/r/defiblockchain/comments/mfzsuy/defichain_node_and_wallet_app_on_raspberry_pi/

**Comment by jha749** [comment]
Here are the instructions to reproduce my proof of concept:

Dockerfile:
```
FROM node:current

WORKDIR /home/node
RUN git clone https://github.com/DefiCh/app

WORKDIR /home/node/app
RUN npm run init

WORKDIR /home/node/app/webapp
#ENTRYPOINT npm run start:react
# The following is just the command associated with start:react with host 0.0.0.0 added, so it will be accessible from outside the docker container
ENTRYPOINT npx cross-env NODE_ENV=development webpack serve --host 0.0.0.0 --config webpack.config.babel.js
EXPOSE 5000
```
Build it using: `docker build -t defi-wallet .`
Then start it with: `docker run -p 5000:5000 defi-wallet`

Note that the container size is huge and could be optimized in many ways, but it is mainly meant to show it can be done.

In order to point the app to your own node, I hardcoded the following changes:

**In webapp/src/utils/rpc-client.ts:**
Comment the lines that initialize rpcauth, rpcconnect and rpcport, and hardcode them as below.
```
//   const { rpcauth, rpcconnect } = state.app.rpcConfig;
//   const rpcport = activeNetwork?.rpcport;
     const rpcconnect = 'IP-ADDRESS-OF-YOUR-NODE';
     const rpcport = '8554';
     const rpcauth = 'rpc-user:rpc-password';
```

**In webapp/src/containers/WalletPage/saga.tsx:**
Hardcode const isWalletCreated to true, comment out the other conditions. This makes sure the app thinks the wallet is created and will show the normal options in the left sidebar, instead of just showing the "create wallet" options.
```
const isWalletCreated = true;
```

Now the big question is, how to improve this hacky solution... :)

**Reply by halfmoonbln** [reply]
Hey good thread,
i am also build my node on a raspberry pi 4 8GB model. Ain is running and now i would like to use the webwallet on raspberry, but i'm struggeling with the rpc conf.

Goal: Own Fullnode on raspberry and i can use the webwallet in my LAN...

**Reply by jha749** [reply]
Coincidentally, I finished yesterday with an all-in-one docker container that allows this. It just requires you to add a RPC auth key to your defi.conf. See my repository here, it also contains (minimal) instructions.
https://github.com/jha749/DefiChain-packager/tree/main/wallet-webapp
Curious to hear whether it works for you and if you have any feedback.

Also planning to prepare docker builds for Raspberry and posting them to Docker Hub so it is more convenient for Pi users to start a node.

**Reply by crypto-r** [reply]
@jha749 Great job. Defichain 1.6.3 runs now on my Raspberry with Raspbian buster. Webwallet 2.3.3. is available on port 5000 in my LAN.

**Comment by Martin8617** [comment]
Rather than the Webapp I liked to use the Defi-Wallet app.
I made a compilation for myself and I would like to serve it to you.

----------------------------------------------------------------------------------------------
Maybe it's possible to add the Raspberry Defi-Wallet to defich/app/releases as well?
Who of the devs can support me to do that?
---------------------------------------------------------------------------------------------

I'd like to upload the AppImage for you, however it's not possible... (even as zip)

**Reply by Martin8617** [reply]
In the meantime I've made my own repositories at github.com/Martin8617/Defi-Wallet-for-Raspberry-Pi
